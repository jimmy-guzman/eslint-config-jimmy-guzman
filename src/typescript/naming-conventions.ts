export const namingConventions = [
  {
    selector: 'default',
    format: ['camelCase'],
  },
  {
    selector: 'variable',
    format: ['camelCase', 'PascalCase', 'UPPER_CASE'],
  },
  {
    selector: 'variable',
    types: ['boolean'],
    format: ['PascalCase'],
    prefix: ['is', 'should', 'has', 'can', 'did', 'will'],
  },
  {
    selector: 'function',
    format: ['camelCase', 'PascalCase'],
  },
  {
    selector: 'parameter',
    format: ['camelCase'],
    leadingUnderscore: 'allow',
  },
  {
    selector: 'typeLike',
    format: ['PascalCase'],
  },
  {
    selector: 'enum',
    format: ['PascalCase', 'UPPER_CASE'],
  },
  {
    selector: 'enumMember',
    format: ['PascalCase', 'UPPER_CASE'],
  },
  {
    selector: 'interface',
    format: ['PascalCase'],
    custom: {
      regex: '^I[A-Z]',
      match: false,
    },
  },
  {
    selector: 'memberLike',
    modifiers: ['private'],
    format: ['camelCase'],
    leadingUnderscore: 'require',
  },
  {
    selector: 'property',
    types: ['boolean', 'string'],
    format: null,
  },
  {
    selector: 'method',
    format: ['camelCase', 'PascalCase'],
  },
]
