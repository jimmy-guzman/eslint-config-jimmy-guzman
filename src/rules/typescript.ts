import type { EslintRules } from '../models'

import { namingConventions } from './naming-conventions'
import { eslintOverrides, importDuplicates } from './typescript-overrides'

/**
 * https://typescript-eslint.io/rules
 */
export const rules: EslintRules = {
  ...importDuplicates,
  ...eslintOverrides,
  '@typescript-eslint/adjacent-overload-signatures': 'error',
  '@typescript-eslint/array-type': 'off', // revisit
  '@typescript-eslint/await-thenable': 'error',
  '@typescript-eslint/ban-ts-comment': 'error',
  '@typescript-eslint/ban-tslint-comment': 'error',
  '@typescript-eslint/ban-types': 'error',
  '@typescript-eslint/block-spacing': 'off',
  '@typescript-eslint/class-literal-property-style': 'off', // revisit,
  '@typescript-eslint/class-methods-use-this': 'off', // revisit
  '@typescript-eslint/consistent-generic-constructors': [
    'error',
    'constructor',
  ],
  '@typescript-eslint/consistent-indexed-object-style': ['error', 'record'],
  '@typescript-eslint/consistent-type-assertions': 'off', // revisit
  '@typescript-eslint/consistent-type-definitions': ['error', 'interface'],
  '@typescript-eslint/consistent-type-exports': [
    'error',
    { fixMixedExportsWithInlineTypeSpecifier: true },
  ],
  '@typescript-eslint/consistent-type-imports': [
    'error',
    { prefer: 'type-imports', disallowTypeAnnotations: true },
  ],
  '@typescript-eslint/default-param-last': 'error',
  '@typescript-eslint/dot-notation': 'error',
  '@typescript-eslint/explicit-function-return-type': 'off',
  '@typescript-eslint/explicit-member-accessibility': 'off', // revisit
  '@typescript-eslint/explicit-module-boundary-types': 'error',
  '@typescript-eslint/init-declarations': 'off',
  '@typescript-eslint/key-spacing': 'off',
  '@typescript-eslint/lines-around-comment': 'off',
  '@typescript-eslint/lines-between-class-members': 'off', // revisit,
  '@typescript-eslint/max-params': ['error', { max: 3 }],
  '@typescript-eslint/member-ordering': [
    'error',
    {
      default: {
        memberTypes: ['signature', 'method', 'constructor', 'field'],
        order: 'alphabetically',
      },
    },
  ],
  '@typescript-eslint/method-signature-style': ['error', 'property'], // fixable
  '@typescript-eslint/naming-convention': ['error', ...namingConventions],
  '@typescript-eslint/no-array-constructor': 'error',
  '@typescript-eslint/no-base-to-string': 'warn',
  '@typescript-eslint/no-confusing-non-null-assertion': 'off', //fixable
  '@typescript-eslint/no-confusing-void-expression': [
    'error',
    { ignoreArrowShorthand: true, ignoreVoidOperator: false },
  ],
  '@typescript-eslint/no-dupe-class-members': 'off',
  '@typescript-eslint/no-duplicate-enum-values': 'error',
  '@typescript-eslint/no-duplicate-type-constituents': 'error',
  '@typescript-eslint/no-dynamic-delete': 'error', // fixable
  '@typescript-eslint/no-empty-function': 'error',
  '@typescript-eslint/no-empty-interface': 'error',
  '@typescript-eslint/no-explicit-any': 'warn',
  '@typescript-eslint/no-extra-non-null-assertion': 'error',
  '@typescript-eslint/no-extra-semi': 'error',
  '@typescript-eslint/no-extraneous-class': 'error',
  '@typescript-eslint/no-floating-promises': 'off', // revisit
  '@typescript-eslint/no-for-in-array': 'error',
  '@typescript-eslint/no-implied-eval': 'error',
  '@typescript-eslint/no-import-type-side-effects': 'error',
  '@typescript-eslint/no-inferrable-types': 'error',
  '@typescript-eslint/no-invalid-this': 'error',
  '@typescript-eslint/no-invalid-void-type': 'warn',
  '@typescript-eslint/no-loop-func': 'error',
  '@typescript-eslint/no-loss-of-precision': 'error',
  '@typescript-eslint/no-magic-numbers': 'off',
  '@typescript-eslint/no-meaningless-void-operator': 'error',
  '@typescript-eslint/no-misused-new': 'error',
  '@typescript-eslint/no-misused-promises': 'warn',
  '@typescript-eslint/no-mixed-enums': 'error',
  '@typescript-eslint/no-namespace': 'error',
  '@typescript-eslint/no-non-null-asserted-nullish-coalescing': 'error',
  '@typescript-eslint/no-non-null-asserted-optional-chain': 'error',
  '@typescript-eslint/no-non-null-assertion': 'warn',
  '@typescript-eslint/no-redeclare': 'off',
  '@typescript-eslint/no-redundant-type-constituents': 'error',
  '@typescript-eslint/no-require-imports': 'off',
  '@typescript-eslint/no-restricted-imports': 'off',
  '@typescript-eslint/no-shadow': 'error',
  '@typescript-eslint/no-this-alias': 'error',
  '@typescript-eslint/no-throw-literal': [
    'error',
    {
      allowThrowingAny: true,
      allowThrowingUnknown: true,
    },
  ],
  '@typescript-eslint/no-type-alias': 'off',
  '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error', // fixable
  '@typescript-eslint/no-unnecessary-condition': 'error', // fixable
  '@typescript-eslint/no-unnecessary-qualifier': 'error', // fixable
  '@typescript-eslint/no-unnecessary-type-arguments': 'error', // fixable
  '@typescript-eslint/no-unnecessary-type-assertion': 'error', // fixable
  '@typescript-eslint/no-unnecessary-type-constraint': 'error', // fixable
  '@typescript-eslint/no-unsafe-argument': 'error',
  '@typescript-eslint/no-unsafe-assignment': 'off', // revisit
  '@typescript-eslint/no-unsafe-call': 'off', // revisit
  '@typescript-eslint/no-unsafe-declaration-merging': 'error',
  '@typescript-eslint/no-unsafe-enum-comparison': 'error',
  '@typescript-eslint/no-unsafe-member-access': 'off', // revisit,
  '@typescript-eslint/no-unsafe-return': 'off', // revisit
  '@typescript-eslint/no-unsafe-unary-minus': 'error',
  '@typescript-eslint/no-unused-expressions': 'error',
  '@typescript-eslint/no-unused-vars': 'warn',
  '@typescript-eslint/no-use-before-define': ['error', 'nofunc'],
  '@typescript-eslint/no-useless-constructor': 'off',
  '@typescript-eslint/no-useless-empty-export': 'error',
  '@typescript-eslint/no-useless-template-literals': 'error',
  '@typescript-eslint/no-var-requires': 'error',
  '@typescript-eslint/non-nullable-type-assertion-style': 'off',
  '@typescript-eslint/padding-line-between-statements': 'off',
  '@typescript-eslint/parameter-properties': 'off', // revisit
  '@typescript-eslint/prefer-as-const': 'error',
  '@typescript-eslint/prefer-destructuring': 'error',
  '@typescript-eslint/prefer-enum-initializers': 'off',
  '@typescript-eslint/prefer-for-of': 'error',
  '@typescript-eslint/prefer-function-type': 'error', // fixable
  '@typescript-eslint/prefer-includes': 'error', // fixable
  '@typescript-eslint/prefer-literal-enum-member': 'error',
  '@typescript-eslint/prefer-namespace-keyword': 'error',
  '@typescript-eslint/prefer-nullish-coalescing': 'error',
  '@typescript-eslint/prefer-optional-chain': 'error',
  '@typescript-eslint/prefer-readonly': 'error', // fixable
  '@typescript-eslint/prefer-readonly-parameter-types': 'off', // revisit
  '@typescript-eslint/prefer-reduce-type-parameter': 'warn', // fixable
  '@typescript-eslint/prefer-regexp-exec': 'error',
  '@typescript-eslint/prefer-return-this-type': 'off', // revisit
  '@typescript-eslint/prefer-string-starts-ends-with': 'error', // fixable
  '@typescript-eslint/prefer-ts-expect-error': 'error', //fixable
  '@typescript-eslint/promise-function-async': 'off',
  '@typescript-eslint/require-array-sort-compare': 'off',
  '@typescript-eslint/require-await': 'error',
  '@typescript-eslint/restrict-plus-operands': 'off',
  '@typescript-eslint/restrict-template-expressions': 'off',
  '@typescript-eslint/return-await': 'error', // fixable
  '@typescript-eslint/sort-type-constituents': 'error',
  '@typescript-eslint/space-before-blocks': 'off',
  '@typescript-eslint/strict-boolean-expressions': 'off', // revisit
  '@typescript-eslint/switch-exhaustiveness-check': 'error',
  '@typescript-eslint/triple-slash-reference': 'error',
  '@typescript-eslint/typedef': 'off', // revisit
  '@typescript-eslint/unbound-method': 'error',
  '@typescript-eslint/unified-signatures': 'warn',
}
