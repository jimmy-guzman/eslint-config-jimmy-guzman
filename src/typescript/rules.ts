import { EslintRules } from '../models'
import { namingConventions } from './naming-conventions'

export const rules: EslintRules = {
  '@typescript-eslint/adjacent-overload-signatures': 'error',
  /**
   * - NOTES: fixable
   */
  '@typescript-eslint/array-type': 'off', // revisit
  /**
   * - TYPE: `üö® error`
   * - RATIONALE: Prevents awaiting values that are not promises which is often an error
   * - FIXABLE: false
   * - NOTES: requires parser settings
   */
  '@typescript-eslint/await-thenable': 'error',
  '@typescript-eslint/ban-ts-comment': 'error',
  '@typescript-eslint/ban-tslint-comment': 'error', // fixable
  '@typescript-eslint/ban-types': 'error',
  /**
   * - NOTES: fixable
   */
  '@typescript-eslint/class-literal-property-style': 'off', // revisit
  '@typescript-eslint/consistent-indexed-object-style': ['error', 'record'], // fixable
  '@typescript-eslint/consistent-type-assertions': 'off', // revisit
  '@typescript-eslint/consistent-type-definitions': ['error', 'interface'], // fixable
  /**
   * [@typescript-eslint/consistent-type-exports](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/consistent-type-exports.md)
   * - TYPE: `üé® stylistic`
   * - RATIONALE: N/A
   */
  '@typescript-eslint/consistent-type-exports': 'off',
  /**
   * - NOTES: fixable
   */
  '@typescript-eslint/consistent-type-imports': 'off', // revisit
  '@typescript-eslint/default-param-last': 'error',
  '@typescript-eslint/dot-notation': 'error', // fixable
  /**
   * [@typescript-eslint/explicit-function-return-type](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/explicit-function-return-type.md)
   * - TYPE: `‚ôªÔ∏è best-practice`
   * - RATIONALE: Explicit types for function return values makes it clear to any calling code what type is returned but this can be annoying and by having `explicit-module-boundary-types` it should suffice
   */
  '@typescript-eslint/explicit-function-return-type': 'off',
  /**
   * - NOTES: fixable
   */
  '@typescript-eslint/explicit-member-accessibility': 'off', // revisit
  /**
   * [@typescript-eslint/explicit-module-boundary-types](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md)
   * - TYPE: `‚ôªÔ∏è best-practice`
   * - RATIONALE: Explicit types for function return values and arguments makes it clear to any calling code what is the module boundary's input and output
   */
  '@typescript-eslint/explicit-module-boundary-types': 'error',
  '@typescript-eslint/init-declarations': 'off',
  /**
   * - NOTES: fixable
   */
  '@typescript-eslint/lines-between-class-members': 'off', // revisit,
  '@typescript-eslint/member-ordering': [
    'error',
    {
      default: {
        memberTypes: ['signature', 'method', 'constructor', 'field'],
        order: 'alphabetically',
      },
    },
  ],
  '@typescript-eslint/method-signature-style': ['error', 'property'], // fixable
  /**
   * [@typescript-eslint/naming-convention](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/naming-convention.md)
   * - TYPE: `üé® stylistic`
   * - RATIONALE: Enforcing naming conventions helps keep the codebase consistent, and reduces overhead when thinking about how to name a variable.
   */
  '@typescript-eslint/naming-convention': ['error', ...namingConventions],
  '@typescript-eslint/no-array-constructor': 'error',
  '@typescript-eslint/no-base-to-string': 'warn',
  '@typescript-eslint/no-confusing-non-null-assertion': 'off', //fixable
  /**
   * [@typescript-eslint/no-confusing-void-expression](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/no-confusing-void-expression.md)
   * - TYPE: `‚ôªÔ∏è best-practice`
   * - RATIONALE: Returning the results of an expression whose type is void can be misleading. `ignoreArrowShorthand` is enabled since it collides with `prettier`
   * - FIXABLE: true
   */
  '@typescript-eslint/no-confusing-void-expression': [
    'error',
    { ignoreArrowShorthand: true, ignoreVoidOperator: false },
  ],
  '@typescript-eslint/no-dupe-class-members': 'off',
  '@typescript-eslint/no-duplicate-imports': 'error',
  '@typescript-eslint/no-dynamic-delete': 'error', // fixable
  '@typescript-eslint/no-empty-function': 'error',
  '@typescript-eslint/no-empty-interface': 'error',
  '@typescript-eslint/no-explicit-any': 'warn',
  '@typescript-eslint/no-extra-non-null-assertion': 'error',
  '@typescript-eslint/no-extra-semi': 'error',
  '@typescript-eslint/no-extraneous-class': 'error',
  '@typescript-eslint/no-floating-promises': 'off', // revisit
  '@typescript-eslint/no-for-in-array': 'error',
  '@typescript-eslint/no-implicit-any-catch': 'error', // fixable
  '@typescript-eslint/no-implied-eval': 'error',
  '@typescript-eslint/no-inferrable-types': 'error',
  '@typescript-eslint/no-invalid-this': 'error',
  '@typescript-eslint/no-invalid-void-type': 'warn',
  '@typescript-eslint/no-loop-func': 'error',
  '@typescript-eslint/no-loss-of-precision': 'error',
  '@typescript-eslint/no-magic-numbers': 'off',
  /**
   * [@typescript-eslint/no-meaningless-void-operator](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/no-meaningless-void-operator.md)
   * - TYPE: `‚ôªÔ∏è best-practices`
   * - RATIONALE: Helps an author catch API changes where previously a value was being discarded at a call site, but the callee changed so it no longer returns a value
   */
  '@typescript-eslint/no-meaningless-void-operator': 'error',
  '@typescript-eslint/no-misused-new': 'error',
  '@typescript-eslint/no-misused-promises': 'warn',
  '@typescript-eslint/no-namespace': 'error',
  /**
   * [@typescript-eslint/no-non-null-asserted-nullish-coalescing](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/no-non-null-asserted-nullish-coalescing.md)
   * - TYPE: `‚ôªÔ∏è best-practices`
   * - RATIONALE: Using non-null assertions in the left operand of the nullish coalescing operator is redundant
   */
  '@typescript-eslint/no-non-null-asserted-nullish-coalescing': 'error',
  '@typescript-eslint/no-non-null-asserted-optional-chain': 'error',
  '@typescript-eslint/no-non-null-assertion': 'warn',
  '@typescript-eslint/no-parameter-properties': 'error',
  '@typescript-eslint/no-redeclare': 'off',
  '@typescript-eslint/no-require-imports': 'off',
  /**
   * [@typescript-eslint/no-restricted-imports]()
   * - TYPE: `üé® stylistic`
   * - RATIONALE: This rule is project dependent
   */
  '@typescript-eslint/no-restricted-imports': 'off',
  '@typescript-eslint/no-shadow': 'error',
  '@typescript-eslint/no-this-alias': 'error',
  /**
   * [no-throw-literal](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/no-throw-literal.md)
   * - TYPE: `‚ôªÔ∏è best-practices`
   * - RATIONALE: It is considered good practice to only throw the `Error` object itself or an object using the `Error` because they automatically keep track of where they were built and originated
   */
  '@typescript-eslint/no-throw-literal': [
    'error',
    {
      allowThrowingAny: true,
      allowThrowingUnknown: true,
    },
  ],
  '@typescript-eslint/no-type-alias': 'off',
  '@typescript-eslint/no-unnecessary-boolean-literal-compare': 'error', // fixable
  '@typescript-eslint/no-unnecessary-condition': 'error', // fixable
  '@typescript-eslint/no-unnecessary-qualifier': 'error', // fixable
  '@typescript-eslint/no-unnecessary-type-arguments': 'error', // fixable
  '@typescript-eslint/no-unnecessary-type-assertion': 'error', // fixable
  '@typescript-eslint/no-unnecessary-type-constraint': 'error', // fixable
  /**
   * [@typescript-eslint/no-unsafe-argument](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/no-unsafe-argument.md)
   * - TYPE: `‚ôªÔ∏è best-practices`
   * - RATIONALE: Calling a function with `any` typed argument can create a potential safety hole, and source of bugs in your codebase
   */
  '@typescript-eslint/no-unsafe-argument': 'error',
  '@typescript-eslint/no-unsafe-assignment': 'off', // revisit
  '@typescript-eslint/no-unsafe-call': 'off', // revisit
  '@typescript-eslint/no-unsafe-member-access': 'off', // revisit
  '@typescript-eslint/no-unsafe-return': 'off', // revisit
  '@typescript-eslint/no-unused-expressions': 'error',
  '@typescript-eslint/no-unused-vars': 'warn',
  '@typescript-eslint/no-use-before-define': ['error', 'nofunc'],
  '@typescript-eslint/no-useless-constructor': 'off',
  '@typescript-eslint/no-var-requires': 'error',
  /**
   * [@typescript-eslint/non-nullable-type-assertion-style](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/non-nullable-type-assertion-style.md)
   * - TYPE: `üé® stylistic`
   * - RATIONALE: Verbose type casts offer more information
   */
  '@typescript-eslint/non-nullable-type-assertion-style': 'off',
  /**
   * [@typescript-eslint/padding-line-between-statement](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/padding-line-between-statements.md)
   * - TYPE: `üé® stylistic`
   * - RATIONALE: padding between types or interface definitions does not improve readability
   */
  '@typescript-eslint/padding-line-between-statements': 'off',
  '@typescript-eslint/prefer-as-const': 'error',
  '@typescript-eslint/prefer-enum-initializers': 'error',
  '@typescript-eslint/prefer-for-of': 'error',
  '@typescript-eslint/prefer-function-type': 'error', // fixable
  '@typescript-eslint/prefer-includes': 'error', // fixable
  '@typescript-eslint/prefer-literal-enum-member': 'error',
  '@typescript-eslint/prefer-namespace-keyword': 'error',
  '@typescript-eslint/prefer-nullish-coalescing': 'error',
  '@typescript-eslint/prefer-optional-chain': 'error',
  '@typescript-eslint/prefer-readonly': 'error', // fixable
  '@typescript-eslint/prefer-readonly-parameter-types': 'off', // revisit
  '@typescript-eslint/prefer-reduce-type-parameter': 'warn', // fixable
  '@typescript-eslint/prefer-regexp-exec': 'error',
  /**
   * [@typescript-eslint/prefer-return-this-type](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/prefer-return-this-type.md)
   * - TYPE: N/A
   * - RATIONALE: N/A
   */
  '@typescript-eslint/prefer-return-this-type': 'off', // revisit
  '@typescript-eslint/prefer-string-starts-ends-with': 'error', // fixable
  '@typescript-eslint/prefer-ts-expect-error': 'error', //fixable
  '@typescript-eslint/promise-function-async': 'off',
  '@typescript-eslint/require-array-sort-compare': 'off',
  '@typescript-eslint/require-await': 'error',
  '@typescript-eslint/restrict-plus-operands': 'off',
  '@typescript-eslint/restrict-template-expressions': 'off',
  '@typescript-eslint/return-await': 'error', // fixable
  /**
   * [@typescript-eslint/sort-type-union-intersection-members](https://github.com/typescript-eslint/typescript-eslint/blob/v5.8.1/packages/eslint-plugin/docs/rules/sort-type-union-intersection-members.md)
   * - TYPE: `üé® stylistic`
   * - RATIONALE: Sorting can help keep your codebase standardized, find repeated types and reduce diff churn
   */
  '@typescript-eslint/sort-type-union-intersection-members': 'error',
  '@typescript-eslint/strict-boolean-expressions': 'off', // revisit
  '@typescript-eslint/switch-exhaustiveness-check': 'error',
  '@typescript-eslint/triple-slash-reference': 'error',
  '@typescript-eslint/typedef': 'off', // revisit
  '@typescript-eslint/unbound-method': 'error',
  '@typescript-eslint/unified-signatures': 'warn',
}
