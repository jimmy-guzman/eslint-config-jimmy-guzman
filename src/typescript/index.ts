import { eslintOverrides, importDuplicates } from './overrides'
import { rules } from './rules'

export = {
  extends: ['prettier'],
  overrides: [
    {
      files: ['**/*.ts?(x)'],
      parser: '@typescript-eslint/parser',
      plugins: ['@typescript-eslint'],
      rules: {
        ...importDuplicates,
        ...eslintOverrides,
        ...rules,
      },
      settings: {
        'import/parsers': {
          '@typescript-eslint/parser': ['.ts', '.tsx', '.d.ts'],
        },
      },
    },
  ],
}
