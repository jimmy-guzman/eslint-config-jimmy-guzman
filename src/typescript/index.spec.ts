import typescript from './index'

describe('typescript', () => {
  it('should export typescript rules config', () => {
    expect(Object.keys(typescript.overrides[0].rules)).toHaveLength(144)
    expect(typescript).toMatchSnapshot()
  })
})
