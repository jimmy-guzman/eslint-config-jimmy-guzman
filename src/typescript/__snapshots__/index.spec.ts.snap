// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typescript should export typescript rules config 1`] = `
Object {
  "extends": Array [
    "prettier",
  ],
  "overrides": Array [
    Object {
      "files": Array [
        "**/*.ts?(x)",
      ],
      "parser": "@typescript-eslint/parser",
      "plugins": Array [
        "@typescript-eslint",
      ],
      "rules": Object {
        "@typescript-eslint/adjacent-overload-signatures": "error",
        "@typescript-eslint/array-type": "off",
        "@typescript-eslint/await-thenable": "error",
        "@typescript-eslint/ban-ts-comment": "error",
        "@typescript-eslint/ban-tslint-comment": "error",
        "@typescript-eslint/ban-types": "error",
        "@typescript-eslint/class-literal-property-style": "off",
        "@typescript-eslint/consistent-indexed-object-style": Array [
          "error",
          "record",
        ],
        "@typescript-eslint/consistent-type-assertions": "off",
        "@typescript-eslint/consistent-type-definitions": Array [
          "error",
          "interface",
        ],
        "@typescript-eslint/consistent-type-exports": "off",
        "@typescript-eslint/consistent-type-imports": "off",
        "@typescript-eslint/default-param-last": "error",
        "@typescript-eslint/dot-notation": "error",
        "@typescript-eslint/explicit-function-return-type": "off",
        "@typescript-eslint/explicit-member-accessibility": "off",
        "@typescript-eslint/explicit-module-boundary-types": "error",
        "@typescript-eslint/init-declarations": "off",
        "@typescript-eslint/lines-between-class-members": "off",
        "@typescript-eslint/member-ordering": Array [
          "error",
          Object {
            "default": Object {
              "memberTypes": Array [
                "signature",
                "method",
                "constructor",
                "field",
              ],
              "order": "alphabetically",
            },
          },
        ],
        "@typescript-eslint/method-signature-style": Array [
          "error",
          "property",
        ],
        "@typescript-eslint/naming-convention": Array [
          "error",
          Object {
            "format": Array [
              "camelCase",
            ],
            "selector": "default",
          },
          Object {
            "format": Array [
              "camelCase",
              "PascalCase",
              "UPPER_CASE",
            ],
            "selector": "variable",
          },
          Object {
            "format": Array [
              "PascalCase",
            ],
            "prefix": Array [
              "is",
              "should",
              "has",
              "can",
              "did",
              "will",
            ],
            "selector": "variable",
            "types": Array [
              "boolean",
            ],
          },
          Object {
            "format": Array [
              "camelCase",
              "PascalCase",
            ],
            "selector": "function",
          },
          Object {
            "format": Array [
              "camelCase",
            ],
            "leadingUnderscore": "allow",
            "selector": "parameter",
          },
          Object {
            "format": Array [
              "PascalCase",
            ],
            "selector": "typeLike",
          },
          Object {
            "format": Array [
              "PascalCase",
              "UPPER_CASE",
            ],
            "selector": "enum",
          },
          Object {
            "format": Array [
              "PascalCase",
              "UPPER_CASE",
            ],
            "selector": "enumMember",
          },
          Object {
            "custom": Object {
              "match": false,
              "regex": "^I[A-Z]",
            },
            "format": Array [
              "PascalCase",
            ],
            "selector": "interface",
          },
          Object {
            "format": Array [
              "camelCase",
            ],
            "leadingUnderscore": "require",
            "modifiers": Array [
              "private",
            ],
            "selector": "memberLike",
          },
          Object {
            "format": null,
            "selector": "property",
          },
          Object {
            "format": Array [
              "camelCase",
              "PascalCase",
            ],
            "selector": "method",
          },
        ],
        "@typescript-eslint/no-array-constructor": "error",
        "@typescript-eslint/no-base-to-string": "warn",
        "@typescript-eslint/no-confusing-non-null-assertion": "off",
        "@typescript-eslint/no-confusing-void-expression": Array [
          "error",
          Object {
            "ignoreArrowShorthand": true,
            "ignoreVoidOperator": false,
          },
        ],
        "@typescript-eslint/no-dupe-class-members": "off",
        "@typescript-eslint/no-duplicate-imports": "error",
        "@typescript-eslint/no-dynamic-delete": "error",
        "@typescript-eslint/no-empty-function": "error",
        "@typescript-eslint/no-empty-interface": "error",
        "@typescript-eslint/no-explicit-any": "warn",
        "@typescript-eslint/no-extra-non-null-assertion": "error",
        "@typescript-eslint/no-extra-semi": "error",
        "@typescript-eslint/no-extraneous-class": "error",
        "@typescript-eslint/no-floating-promises": "off",
        "@typescript-eslint/no-for-in-array": "error",
        "@typescript-eslint/no-implicit-any-catch": "error",
        "@typescript-eslint/no-implied-eval": "error",
        "@typescript-eslint/no-inferrable-types": "error",
        "@typescript-eslint/no-invalid-this": "error",
        "@typescript-eslint/no-invalid-void-type": "warn",
        "@typescript-eslint/no-loop-func": "error",
        "@typescript-eslint/no-loss-of-precision": "error",
        "@typescript-eslint/no-magic-numbers": "off",
        "@typescript-eslint/no-meaningless-void-operator": "error",
        "@typescript-eslint/no-misused-new": "error",
        "@typescript-eslint/no-misused-promises": "warn",
        "@typescript-eslint/no-namespace": "error",
        "@typescript-eslint/no-non-null-asserted-nullish-coalescing": "error",
        "@typescript-eslint/no-non-null-asserted-optional-chain": "error",
        "@typescript-eslint/no-non-null-assertion": "warn",
        "@typescript-eslint/no-parameter-properties": "error",
        "@typescript-eslint/no-redeclare": "off",
        "@typescript-eslint/no-redundant-type-constituents": "error",
        "@typescript-eslint/no-require-imports": "off",
        "@typescript-eslint/no-restricted-imports": "off",
        "@typescript-eslint/no-shadow": "error",
        "@typescript-eslint/no-this-alias": "error",
        "@typescript-eslint/no-throw-literal": Array [
          "error",
          Object {
            "allowThrowingAny": true,
            "allowThrowingUnknown": true,
          },
        ],
        "@typescript-eslint/no-type-alias": "off",
        "@typescript-eslint/no-unnecessary-boolean-literal-compare": "error",
        "@typescript-eslint/no-unnecessary-condition": "error",
        "@typescript-eslint/no-unnecessary-qualifier": "error",
        "@typescript-eslint/no-unnecessary-type-arguments": "error",
        "@typescript-eslint/no-unnecessary-type-assertion": "error",
        "@typescript-eslint/no-unnecessary-type-constraint": "error",
        "@typescript-eslint/no-unsafe-argument": "error",
        "@typescript-eslint/no-unsafe-assignment": "off",
        "@typescript-eslint/no-unsafe-call": "off",
        "@typescript-eslint/no-unsafe-member-access": "off",
        "@typescript-eslint/no-unsafe-return": "off",
        "@typescript-eslint/no-unused-expressions": "error",
        "@typescript-eslint/no-unused-vars": "warn",
        "@typescript-eslint/no-use-before-define": Array [
          "error",
          "nofunc",
        ],
        "@typescript-eslint/no-useless-constructor": "off",
        "@typescript-eslint/no-useless-empty-export": "error",
        "@typescript-eslint/no-var-requires": "error",
        "@typescript-eslint/non-nullable-type-assertion-style": "off",
        "@typescript-eslint/padding-line-between-statements": "off",
        "@typescript-eslint/prefer-as-const": "error",
        "@typescript-eslint/prefer-enum-initializers": "error",
        "@typescript-eslint/prefer-for-of": "error",
        "@typescript-eslint/prefer-function-type": "error",
        "@typescript-eslint/prefer-includes": "error",
        "@typescript-eslint/prefer-literal-enum-member": "error",
        "@typescript-eslint/prefer-namespace-keyword": "error",
        "@typescript-eslint/prefer-nullish-coalescing": "error",
        "@typescript-eslint/prefer-optional-chain": "error",
        "@typescript-eslint/prefer-readonly": "error",
        "@typescript-eslint/prefer-readonly-parameter-types": "off",
        "@typescript-eslint/prefer-reduce-type-parameter": "warn",
        "@typescript-eslint/prefer-regexp-exec": "error",
        "@typescript-eslint/prefer-return-this-type": "off",
        "@typescript-eslint/prefer-string-starts-ends-with": "error",
        "@typescript-eslint/prefer-ts-expect-error": "error",
        "@typescript-eslint/promise-function-async": "off",
        "@typescript-eslint/require-array-sort-compare": "off",
        "@typescript-eslint/require-await": "error",
        "@typescript-eslint/restrict-plus-operands": "off",
        "@typescript-eslint/restrict-template-expressions": "off",
        "@typescript-eslint/return-await": "error",
        "@typescript-eslint/sort-type-union-intersection-members": "error",
        "@typescript-eslint/space-before-blocks": "off",
        "@typescript-eslint/strict-boolean-expressions": "off",
        "@typescript-eslint/switch-exhaustiveness-check": "error",
        "@typescript-eslint/triple-slash-reference": "error",
        "@typescript-eslint/typedef": "off",
        "@typescript-eslint/unbound-method": "error",
        "@typescript-eslint/unified-signatures": "warn",
        "constructor-super": "off",
        "getter-return": "off",
        "import/no-duplicates": "off",
        "no-array-constructor": "off",
        "no-const-assign": "off",
        "no-dupe-args": "off",
        "no-dupe-class-members": "off",
        "no-dupe-keys": "off",
        "no-duplicate-imports": "off",
        "no-empty-function": "off",
        "no-extra-semi": "off",
        "no-func-assign": "off",
        "no-implied-eval": "off",
        "no-import-assign": "off",
        "no-loss-of-precision": "off",
        "no-new-symbol": "off",
        "no-obj-calls": "off",
        "no-redeclare": "off",
        "no-setter-return": "off",
        "no-shadow": "off",
        "no-this-before-super": "off",
        "no-undef": "off",
        "no-unreachable": "off",
        "no-unsafe-negation": "off",
        "no-unused-vars": "off",
        "no-var": "error",
        "prefer-const": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
        "require-await": "off",
        "valid-typeof": "off",
      },
      "settings": Object {
        "import/parsers": Object {
          "@typescript-eslint/parser": Array [
            ".ts",
            ".tsx",
            ".d.ts",
          ],
        },
      },
    },
  ],
}
`;
